### PR TITLE
Add a local_snapshot attribute to stack_snapshot

### LINKED ISSUE #984: Custom snapshots in stack_snapshot
I'd like for stack snapshot to be extensible:

- it is frequent that some dependencies in a stackage snapshot is missing or you need a specific version.
- As far as I can tell, writing Haskell dependencies which are not in the stackage snapshot is pretty painful currently.

What I'm proposing is to make it possible, for `stack_snapshot` to take a custom `snapshot.yaml` (in the stack syntax) rather than a globally named snapshot. This would still let stack do all the dependency solving, with all the convenience and simplicity that it buys us.

It sounds easy enough in principle. But I'm not sure what tools Bazel gives me to do this. So this is very much a design conversation.

- I think that my preferred way would be to make the `snapshot` attribute be either a string or a label, and call something like `--resolver ($location snapshot)` when it is a label. But I don't know if Bazel lets me distinguish between the two.
- The other solution which I can think of at the moment is to have two attributes, one for global snapshot, and one for local snapshots, the global one is a string, the local one is a label, we disambiguate depending on which is set. This looks more straightforward, but feels kind of clunky.

In either case, the label would refer to some `local_stack_snapshot` target, which we would need to add to pull the custom snapshot in scope.

What do you guys think?

PS: I did check that you could pass a custom snapshot in `stack --resolver <snapshot>`.

### PR DESCRIPTION
**Pull request by aspiwack**
It lets one use a custom snapshot file (in the stack syntax) instead of a globally defined one.

A `stack_snapshot` target must have exactly one of `snapshot` or `local_snapshot`.

Fixes #984

### CODE REVIEW

File: haskell/cabal.bzl [hunk 1]
```diff
@@ -635,7 +646,11 @@ stack_snapshot = repository_rule(
     _stack_snapshot_impl,
     attrs = {
         "snapshot": attr.string(
-            doc = "The name of a Stackage snapshot.",
+            doc = "The name of a Stackage snapshot. Incompatible with local_snapshot.",
+        ),
+        "local_snapshot": attr.label(
+            doc = "A custom Stack snapshot file, as per the Stack documentation. Incompatible with snapshot.",
```
Comment (aspiwack) [inline haskell/cabal.bzl]: This PR is a draft because I need to add example of this new attribute before merging. But it's already reviewable.

File: haskell/cabal.bzl [hunk 2]
```diff
@@ -552,6 +552,16 @@ Specify a fully qualified package name of the form <package>-<version>.
...
 def _stack_snapshot_impl(repository_ctx):
+
+    if repository_ctx.attr.snapshot and repository_ctx.attr.local_snapshot:
+        fail("Please specify either snapshot or repository_snapshot, not both")
+    elif repository_ctx.attr.snapshot:
+        snapshot = repository_ctx.attr.snapshot
+    elif repository_ctx.attr.local_snapshot:
+        snapshot = repository_ctx.path(repository_ctx.attr.local_snapshot)
+    else:
+        fail("Please specify one of snapshot or repository_snapshot")
```
Comment (aspiwack) [inline haskell/cabal.bzl]: It doesn't look related to this PR, directly, but for some reason, errors in `stack_snapshot` seem to not stop the execution right away, and you get additional spurious errors. Example with a typo in the name of the snapshot:

```
ERROR: An error occurred during the fetch of repository 'stackage':
   Not a regular file: /home/aspiwack/projects/tweag/spot/master/snapshot.yam
ERROR: /home/aspiwack/projects/tweag/spot/master/backend/server/BUILD.bazel:15:1: no such package '@stackage//': Not a regular file: /home/aspiwack/projects/tweag/spot/master/snapshot.yam and referenced by '//backend/server:server'
ERROR: /home/aspiwack/projects/tweag/spot/master/backend/server/BUILD.bazel:15:1: no such package '@stackage//': Not a regular file: /home/aspiwack/projects/tweag/spot/master/snapshot.yam and referenced by '//backend/server:server'
ERROR: /home/aspiwack/projects/tweag/spot/master/backend/server/BUILD.bazel:15:1: no such package '@stackage//': Not a regular file: /home/aspiwack/projects/tweag/spot/master/snapshot.yam and referenced by '//backend/server:server'
ERROR: /home/aspiwack/projects/tweag/spot/master/backend/server/BUILD.bazel:15:1: no such package '@stackage//': Not a regular file: /home/aspiwack/projects/tweag/spot/master/snapshot.yam and referenced by '//backend/server:server'
ERROR: /home/aspiwack/projects/tweag/spot/master/backend/server/BUILD.bazel:15:1: no such package '@stackage//': Not a regular file: /home/aspiwack/projects/tweag/spot/master/snapshot.yam and referenced by '//backend/server:server'
ERROR: /home/aspiwack/projects/tweag/spot/master/backend/server/BUILD.bazel:15:1: no such package '@stackage//': Not a regular file: /home/aspiwack/projects/tweag/spot/master/snapshot.yam and referenced by '//backend/server:server'
ERROR: /home/aspiwack/projects/tweag/spot/master/backend/server/BUILD.bazel:15:1: no such package '@stackage//': Not a regular file: /home/aspiwack/projects/tweag/spot/master/snapshot.yam and referenced by '//backend/server:server'
ERROR: /home/aspiwack/projects/tweag/spot/master/backend/server/BUILD.bazel:15:1: no such package '@stackage//': Not a regular file: /home/aspiwack/projects/tweag/spot/master/snapshot.yam and referenced by '//backend/server:server'
ERROR: /home/aspiwack/projects/tweag/spot/master/backend/server/BUILD.bazel:15:1: no such package '@stackage//': Not a regular file: /home/aspiwack/projects/tweag/spot/master/snapshot.yam and referenced by '//backend/server:server'
ERROR: /home/aspiwack/projects/tweag/spot/master/backend/server/BUILD.bazel:15:1: no such package '@stackage//': Not a regular file: /home/aspiwack/projects/tweag/spot/master/snapshot.yam and referenced by '//backend/server:server'
ERROR: Analysis of target '//backend/server:server' failed; build aborted: no such package '@stackage//': Not a regular file: /home/aspiwack/projects/tweag/spot/master/snapshot.yam
```

File: haskell/cabal.bzl [hunk 3]
```diff
@@ -552,6 +552,16 @@ Specify a fully qualified package name of the form <package>-<version>.
     return (dependencies, transitive_unpacked_sdists)
 
 def _stack_snapshot_impl(repository_ctx):
+
+    if repository_ctx.attr.snapshot and repository_ctx.attr.local_snapshot:
+        fail("Please specify either snapshot or repository_snapshot, not both")
```
Comment (mboes) [inline haskell/cabal.bzl]:
```suggestion
        fail("Please specify either snapshot or local_snapshot, but not both.")
```

**Comment by aspiwack** [comment]
Latest rebase should fix formatting. I added an example. Ready to merge on my end.
